File: replay/awaker_replay/teacher_rule.py
"""teacher 룰 v0의 replay 재구현 — 앱 `TeacherRule.kt`와 의도적 1:1 복제.

한쪽을 바꾸면 반드시 같이 바꿀 것. debounce·윈도우·임계는 물론 median 인덱싱
(`sorted_gaps[n // 2]` — 상위 중앙값)까지 동일해야 재현성 검증이 성립한다.
시간 단위: 단조 ms (elapsedRealtime).
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass


@dataclass(frozen=True)
class Config:
    window_ms: int = 60_000
    debounce_ms: int = 500
    min_flings: int = 8
    min_span_ms: int = 30_000
    max_median_gap_ms: int = 8_000
    max_entry_pause_ms: int = 20_000
    exit_silence_ms: int = 30_000
    exit_min_flings: int = 4


@dataclass(frozen=True)
class Metrics:
    flings: int
    span_ms: int
    median_gap_ms: int
    max_gap_ms: int


@dataclass(frozen=True)
class Transition:
    at_ms: int
    state: str  # "enter" | "exit"
    metrics: Metrics
    reason: str | None = None


class TeacherRule:
    def __init__(self, config: Config = Config()):
        self.config = config
        self._flings: deque[int] = deque()
        self._last_scroll_ms: int | None = None
        self.is_positive = False

    def on_scroll(self, at_ms: int) -> Transition | None:
        if self._last_scroll_ms is None or at_ms - self._last_scroll_ms >= self.config.debounce_ms:
            self._flings.append(at_ms)
        self._last_scroll_ms = at_ms
        return self._evaluate(at_ms)

    def on_tick(self, at_ms: int) -> Transition | None:
        return self._evaluate(at_ms)

    def reset(self, at_ms: int) -> Transition | None:
        was_positive = self.is_positive
        metrics = self._snapshot()
        self._flings.clear()
        self._last_scroll_ms = None
        self.is_positive = False
        if was_positive:
            return Transition(at_ms, "exit", metrics, reason="reset")
        return None

    def _evaluate(self, now_ms: int) -> Transition | None:
        cfg = self.config
        while self._flings and now_ms - self._flings[0] > cfg.window_ms:
            self._flings.popleft()
        metrics = self._snapshot()

        if not self.is_positive:
            enter = (
                metrics.flings >= cfg.min_flings
                and metrics.span_ms >= cfg.min_span_ms
                and metrics.median_gap_ms <= cfg.max_median_gap_ms
                and metrics.max_gap_ms <= cfg.max_entry_pause_ms
            )
            if enter:
                self.is_positive = True
                return Transition(now_ms, "enter", metrics)
            return None

        silence = (now_ms - self._flings[-1]) if self._flings else float("inf")
        if silence > cfg.exit_silence_ms:
            reason = "silence"
        elif metrics.flings < cfg.exit_min_flings:
            reason = "cadence_collapse"
        else:
            return None
        self.is_positive = False
        return Transition(now_ms, "exit", metrics, reason)

    def _snapshot(self) -> Metrics:
        flings = list(self._flings)
        if len(flings) < 2:
            return Metrics(len(flings), 0, 0, 0)
        gaps = sorted(flings[i + 1] - flings[i] for i in range(len(flings) - 1))
        return Metrics(
            flings=len(flings),
            span_ms=flings[-1] - flings[0],
            median_gap_ms=gaps[len(gaps) // 2],
            max_gap_ms=gaps[-1],
        )

File: app/src/main/kotlin/com/awaker/detection/TeacherRule.kt
package com.awaker.detection

import java.util.ArrayDeque

/**
 * Teacher 룰 v0 — AS 스크롤 운동학 기반 무지성 스크롤 판정 (순수 Kotlin).
 * ADR-0010의 dense 라벨원이자 ADR-0011의 프로토타입 라이브 트리거.
 *
 * 판정 직관(CONTEXT.md 예시 대화): "2초 간격으로 반사적으로 넘기고 멈춤이
 * 없으면 무지성. 속도 변화가 있고 멈춰서 읽는 구간이 있으면 아님."
 *
 * v0 수식 (초안 — 임계는 본인 데이터 self-annotation으로 1차 튜닝, [Config]):
 *
 * fling = 스크롤 이벤트를 debounceMs로 뭉친 제스처. 최근 windowMs 안의 fling
 * 타임스탬프 f₁…fₙ, 간격 gᵢ = fᵢ₊₁ − fᵢ 에 대해
 *
 * 양성 진입 (모두 충족):
 *   n ≥ minFlings                    — 표본 충분
 *   span = fₙ − f₁ ≥ minSpanMs      — 지속된 행동 (순간 폭주 배제)
 *   median(g) ≤ maxMedianGapMs      — 반사적 케이던스
 *   max(g) ≤ maxEntryPauseMs        — 의미 있는 멈춤 없음
 *
 * 양성 해제 (하나라도):
 *   now − fₙ > exitSilenceMs        — 진짜 멈춤 (읽기/시청/이탈)
 *   window 내 n < exitMinFlings     — 케이던스 붕괴
 *   reset() — 세션 종료 (로그에는 세션 end가 곧 암묵적 해제)
 *
 * 시간은 단조 ms(elapsedRealtime). 판정에 X(권한-경량 센서)가 아닌 AS 신호를
 * 쓰는 것은 베타 한정 — 출시 경로에서 제거(ADR-0004).
 */
class TeacherRule(private val config: Config = Config()) {

    data class Config(
        val windowMs: Long = 60_000,
        val debounceMs: Long = 500,
        val minFlings: Int = 8,
        val minSpanMs: Long = 30_000,
        val maxMedianGapMs: Long = 8_000,
        val maxEntryPauseMs: Long = 20_000,
        val exitSilenceMs: Long = 30_000,
        val exitMinFlings: Int = 4,
    )

    /** 판정 시점의 윈도우 운동학 스냅샷 — rule 로그 라인과 replay 대조에 쓰인다. */
    data class Metrics(
        val flings: Int,
        val spanMs: Long,
        val medianGapMs: Long,
        val maxGapMs: Long,
    )

    sealed interface Transition {
        val atMs: Long
        val metrics: Metrics

        data class Enter(override val atMs: Long, override val metrics: Metrics) : Transition
        data class Exit(
            override val atMs: Long,
            override val metrics: Metrics,
            val reason: String,
        ) : Transition
    }

    private val flings = ArrayDeque<Long>()
    private var lastScrollMs = Long.MIN_VALUE

    var isPositive: Boolean = false
        private set

    /** AS 스크롤 이벤트 (debounce 전 raw). 전이가 생기면 반환. */
    fun onScroll(atMs: Long): Transition? {
        if (lastScrollMs == Long.MIN_VALUE || atMs - lastScrollMs >= config.debounceMs) {
            flings.addLast(atMs)
        }
        lastScrollMs = atMs
        return evaluate(atMs)
    }

    /** 주기 평가 — 침묵에 의한 해제는 스크롤 없이도 판정돼야 한다. */
    fun onTick(atMs: Long): Transition? = evaluate(atMs)

    /** 세션 종료 등 외부 리셋. 양성이었다면 Exit을 반환한다. */
    fun reset(atMs: Long): Transition? {
        val wasPositive = isPositive
        val metrics = snapshot()
        flings.clear()
        lastScrollMs = Long.MIN_VALUE
        isPositive = false
        return if (wasPositive) Transition.Exit(atMs, metrics, reason = "reset") else null
    }

    private fun evaluate(nowMs: Long): Transition? {
        while (flings.isNotEmpty() && nowMs - flings.first() > config.windowMs) flings.removeFirst()
        val metrics = snapshot()

        if (!isPositive) {
            val enter = metrics.flings >= config.minFlings &&
                metrics.spanMs >= config.minSpanMs &&
                metrics.medianGapMs <= config.maxMedianGapMs &&
                metrics.maxGapMs <= config.maxEntryPauseMs
            if (enter) {
                isPositive = true
                return Transition.Enter(nowMs, metrics)
            }
            return null
        }

        val silence = if (flings.isEmpty()) Long.MAX_VALUE else nowMs - flings.last()
        val reason = when {
            silence > config.exitSilenceMs -> "silence"
            metrics.flings < config.exitMinFlings -> "cadence_collapse"
            else -> null
        }
        if (reason != null) {
            isPositive = false
            return Transition.Exit(nowMs, metrics, reason)
        }
        return null
    }

    private fun snapshot(): Metrics {
        val list = flings.toLongArray()
        if (list.size < 2) return Metrics(list.size, 0, 0, 0)
        val gaps = LongArray(list.size - 1) { list[it + 1] - list[it] }
        gaps.sort()
        val median = gaps[gaps.size / 2]
        return Metrics(
            flings = list.size,
            spanMs = list.last() - list.first(),
            medianGapMs = median,
            maxGapMs = gaps.last(),
        )
    }
}
